### DISCUSSION TITLE
camera picture for regions of interests is tilted against reference-image

### DISCUSSION THREAD

**Discussion by Harold-3235464** [jomjol/AI-on-the-edge-device#2718] [General]
### The Problem

I am new here, so excuse if problem already has been solved, but I didnt't find it here.
The camera-picture on which the regions of interest are applied, is tilted against the prevousliy taken reference-image. Therefor the reading fails.
I repeated the reference-image-procedure already several times and same with regions of interest, but every time the reading picture is tilted after that.
It would be great to hear, what I am doing wrong.
Pictures to explain the topic:
![version 231126](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/f5060a7a-1073-48bd-a150-dbb9cc601616)
[
![reference image 231126](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/bb424c9e-1d7e-4926-b0b9-5b121e73e303)
![recognition 231126](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/2144f52b-10ee-4288-83df-2b08e434e6bf)
![digital region of interest 231126](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/1866f8c4-9041-430d-a6dd-6c118ea7c77a)
![analog region of interest 231126](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/0710f793-14ec-4cfc-9f47-de3a30cd3b30)
](url)

If you need more data to support, please let me know

### Version

Firmware version: 	Release: v15.3.0 (Commit: 3fbff0a) Firmware build time: 	22.07.2023 09:42 Web interface version: 	Release: v15.3.0 (Commit: 3fbff0a)

### Logfile

```shell
.
```

### Expected Behavior

the camera-picture used for reading should be "same" orientation/angle as reference- and learnt ROIs

### Screenshots

screenshots are appended in problem

### Additional Context

.

**Comment by caco3** [comment]
It is possible that the reference data did not get saved to the SD card properly.

First tip would be to save the reference image and marks once mor3e, but you say you did this already.
You should set the log to DEBUG for those steps and try to really save and even restart between each step.

 Also you could try to use a different SD-Card or check it got inserted properly.

**Reply by Harold-3235464** [reply]
Hello thanks for your response :),
I checked the SD-Card. Reinserted it. Then I made a new reference image and positioned the digital and analog reading-frames by try and error to the faulty position (not aligned to the picture but to the, so the watermeter was reading. Yesterday 29.08, this has been working for a while, but then the tilt changed. All the time, the life-stream does not change the angle. 
Today, 30.09 the original tilt is back againg - without having changed anytining and after that back to the "wrong" tilt.
![2023-11-29 09 08 07](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/50818f87-8ff5-417a-b3b1-fe1a7831b579)
![2023-11-29 09 08 31](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/bfd5af84-4d47-4483-8856-ec95f4452bbb)
![2023-11-30 07 27 17](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/819da8bb-0313-4f33-9327-95004e9ecf86)
![2023-11-30 07 27 45](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/48f48291-bd6b-4db5-b47f-4c2c3571d24e)
![2023-11-30 07 54 01](https://github.com/jomjol/AI-on-the-edge-device/assets/96665050/88d30c77-b37f-4499-b932-1f22ef13c004)

[log_2023-11-29.txt](https://github.com/jomjol/AI-on-the-edge-device/files/13509312/log_2023-11-29.txt)
[log_2023-11-30.txt](https://github.com/jomjol/AI-on-the-edge-device/files/13509467/log_2023-11-30.txt)
[data_2023-11-30.csv](https://github.com/jomjol/AI-on-the-edge-device/files/13509521/data_2023-11-30.csv)

[watermeter_2023-11-30_06-48-40.zip](https://github.com/jomjol/AI-on-the-edge-device/files/13509434/watermeter_2023-11-30_06-48-40.zip)

Hoping the data will help to find out why the tilt changes.
Harold

**Reply by tronauts** [reply]
![image](https://github.com/jomjol/AI-on-the-edge-device/assets/82866068/6689b043-a669-43cd-b723-d0df632d41c3)
I have the same problem.

**Comment by friedpa** [comment]
The alignment marks are not good. Just stick a black dot on the watermeter at the bottom and take the m3 on the top. Secondly correct the lens so that the picture is sharper (your alignment mark on the left side has way to less contrast)

**Comment by Harold-3235464** [comment]
Hello,
thanks for your advice. I will try to follow your instructions. Instead of
M22 in the frame, I will choose the m³ and the dot. The dot is (originally)
chosen.

I will keep you updated about the result. :)
Harold Ebeling
Mohnweg 4
91578 Leutershausen


Am Do., 30. Nov. 2023 um 15:29 Uhr schrieb Paul Fried <
***@***.***>:

**Comment by caco3** [comment]
Also make sure that:
- The reference marks are not too large (increased alignment time)
- The reference marks are as far apart as possible
- The reference marks do not cover any changing parts (digit, pointer)

**Comment by Harold-3235464** [comment]
Hello Paul,

thanks for your advice. This really helped. Now the picture is stable. :)

But may be you may also have an advice for my second problem. I have a
meter where the digits already will show the new digital value, but the
analog meters still haven't finished their circles. So it always comes to
misreadings.
I 've been fighting with the last digit, and thought, that I had solved
it,  but now the second last digital number is changing and I have again
the problem.
As you can see, it now calculates with 100 instead of 109.  So last digit
again is switching too early to 0, but second does not follow (in that case
it should be 1  => 110).
How do I have to setup the configuration (6.x nor 9.9 does help in
analog/digital transition.)?

[image: grafik.png]

thanks in advance for your support

Harold Ebeling
Mohnweg 4
91578 Leutershausen


Am Do., 30. Nov. 2023 um 15:29 Uhr schrieb Paul Fried <
***@***.***>:

**Comment by friedpa** [comment]
Entschuldigung für meine späte Antwort, ich sehe leider das Bild nicht.....

**Comment by Harold-3235464** [comment]
Hallo Paul?,
nochmals danke für die Unterstützung. Ich hatte das Bild direkt in die Mail
kopiert. Jetzt ein erneuter Versuch mit einem aktuelleren Bild (gleiches
Problem) als Anhang.
In dem Bild siehst du einen Zähler 109.790.    Ermittelt wird aber 110.79.

LG

Harold Ebeling
Mohnweg 4
91578 Leutershausen


Am Mi., 6. Dez. 2023 um 09:55 Uhr schrieb Paul Fried <
***@***.***>:

**Reply by Marvel912** [reply]
I had same issues, even tried to upload an older version, updated region of interest and many more, but without success. finally get it working. i have adjusted both allignment marker to totally different position. Now it is working for me *love it*

**Comment by jasaw** [comment]
I have occasional alignment issue and it looks like the image is rotated a bit too much. Looking at the set up process in its entirety, the rotation should be limited to less than a couple of degrees (probably make it configurable).

The reason being:
The camera is mounted as the first step, then a reference image is taken. If the camera is moved, a new reference image has to be taken for things to work optimally anyway. Taking a new reference image requires new alignment markers to be taken. From this point on, the images that are taken shouldn't need a lot of rotation. If the algorithm thinks that it needs 30 degrees rotation, that's obviously wrong, so it makes sense to limit rotation and/or give up on rotation if it can't find the alignment markers.

**Reply by caco3** [reply]
a limiter might be useful, but ir also is not expected for the camera to move after setup!

**Reply by caco3** [reply]
> If the algorithm thinks that it needs 30 degrees rotation, that's obviously wrong, so it makes sense to limit rotation and/or give up on rotation

How much is the rotation off compared to the alignment image?

**Reply by jasaw** [reply]
>How much is the rotation off compared to the alignment image?

I didn't measure it, but I estimate maybe 5 to 10 degrees off, which is enough to throw it off.

If the expectation is the camera shall not move after setup, then it would be better to have a configuration option to turn off alignment altogether (probably by setting max rotation config to zero).

**Reply by Slider0007** [reply]
> a configuration option to turn off

There is already an expert parameter `Alignment Algorithm` implemented to switch off alignment marker mapping algorithm. If you switch this parameter to off only inital rotation (defined in reference image page) will be applied for correction, but no additional pattern mapping with alignment marker.

Same topic: https://github.com/jomjol/AI-on-the-edge-device/discussions/2736#discussioncomment-7821984

**Reply by jasaw** [reply]
Thank you @Slider0007 . I didn't know the auto alignment can be turned off. It's working better now that it's turned off.

**Comment by jasaw** [comment]
@Slider0007 @SybexX @caco3 
The issue with my installation is occasionally I get slight vertical camera movement because of how I mounted my camera. There is absolutely no rotation. If I just use the existing alignment algorithm, it finds the markers without any problem, but it fails badly at the rotation step and ended up rotating my image 30 degrees.

I've added 2 additional alignment algorithm options, shift-only-high-accuracy and shift-only-fast. Both are exactly the same as the existing algorithms except they skip the rotation step and this is working quite well for me from my short amount of testing. This obviously requires more testing and possibly more image enhancement could be done, but this is already a huge improvement compared to turning off alignment.

I've created a PR to get your thoughts on adding this functionality to the code base.
https://github.com/jasaw/AI-on-the-edge-device/pull/1

Question about the alignment algorithm:
The `CalculateSimularities` function finds the minDif, maxDif, average by comparing individual colours. While scanning and searching for the best match, a good match (small dif) on a single colour will result in minDif being recorded while ignoring bad matches on the other 2 colours of the same pixel?
```
            for (_ch = 0; _ch < channels; ++_ch)
            {
                dif = p_tpl[_ch] - p_org[_ch];
                aktSAD += pow(p_tpl[_ch] - p_org[_ch], 2);
                if (dif < minDif) minDif = dif;
                if (dif > maxDif) maxDif = dif;
                avgDifSum += dif;
                anz++;
            }
```

Isn't it more correct to combine the 3 colours that make up a pixel before calculating the dif, min max avg?
```
            if (channels == 3)
            {
                double greyscale_tpl = 0.2989 * p_tpl[0] + 0.5870 * p_tpl[1] + 0.1140 * p_tpl[2];
                double greyscale_org = 0.2989 * p_org[0] + 0.5870 * p_org[1] + 0.1140 * p_org[2];
                dif = greyscale_tpl - greyscale_org;
                aktSAD += dif*dif;
                if (dif < minDif) minDif = dif;
                if (dif > maxDif) maxDif = dif;
                avgDifSum += dif;
                anz++;
            }
```

**Comment by jasaw** [comment]
I've been experimenting with a different alignment algorithm. Here's my findings. Keen to get your thoughts on it.
- The existing alignment algorithm attempts to do a "fast" alignment which normally fails badly, then falls back to "high accuracy" alignment algorithm that matches the template against every single pixel in the search area. The "high accuracy" alignment does a pretty good job but it is VERY slow.
- The "high accuracy" alignment does RGB comparison and it is susceptible to changes caused by auto white balance of the camera, and somewhat susceptible to the auto exposure level. It is also very slow.
- The performance can be increased by combining the RGB into 1 grey pixel, and only compare the grey pixel. This also gets rid of the image colour variation caused by auto white balance.
- The best match is done by finding the lowest SAD value. Performance can be further increased by aborting the computational loop on the current position if the current SAD is already higher than the best match that we have found.
- The brightness variation issue can be addressed by normalizing the brightness of the region of interest (ROI) in the image. Normalizing operation is not too computationally expensive. Find the min and max brightness of the ROI, then apply a scale to the ROI pixels such that the brightness in the ROI area ranges from 0 to 255.

I have implemented this and early testing shows that it is more robust and faster than the "high accuracy" algorithm.
